crates/biome_html_formatter/src/html/lists/element_list.rs

```diff
@@ -1,10 +1,720 @@
-use crate::prelude::*;
-use biome_html_syntax::HtmlElementList;
+//! This implementation is very heavily inspired by the JSX formatter implementation for JsxChildList.
+
+use std::cell::RefCell;
+
+use crate::{
+    comments::HtmlComments,
+    prelude::*,
+    utils::children::{
+        html_split_children, is_meaningful_html_text, HtmlChild, HtmlChildrenIterator,
+        HtmlRawSpace, HtmlSpace,
+    },
+};
+use biome_formatter::{best_fitting, prelude::*, CstFormatContext};
+use biome_formatter::{format_args, write, VecBuffer};
+use biome_html_syntax::{AnyHtmlElement, HtmlElementList};
+use tag::GroupMode;
 #[derive(Debug, Clone, Default)]
-pub(crate) struct FormatHtmlElementList;
+pub(crate) struct FormatHtmlElementList {
+    layout: HtmlChildListLayout,
+}
 impl FormatRule<HtmlElementList> for FormatHtmlElementList {
     type Context = HtmlFormatContext;
     fn fmt(&self, node: &HtmlElementList, f: &mut HtmlFormatter) -> FormatResult<()> {
-        f.join().entries(node.iter().formatted()).finish()
+        if node.is_empty() {
+            return Ok(());
+        }
+
+        let result = self.fmt_children(node, f)?;
+        match result {
+            FormatChildrenResult::ForceMultiline(format_multiline) => {
+                write!(f, [format_multiline])
+            }
+            FormatChildrenResult::BestFitting {
+                flat_children,
+                expanded_children,
+            } => {
+                write!(f, [best_fitting![flat_children, expanded_children]])
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum FormatChildrenResult {
+    ForceMultiline(FormatMultilineChildren),
+    BestFitting {
+        flat_children: FormatFlatChildren,
+        expanded_children: FormatMultilineChildren,
+    },
+}
```

Comment:
Should we name this `FormatChildrenLayout` instead? We should also document the variants and explain when they are applied, maybe with a code snippet

Reply:
Actually, upon further inspection, we already have the `HtmlChildListLayout` enum, so renaming this to `FormatChildrenLayout` could get confusing.